packages/gatsby/src/utils/map-pages-to-static-query-hashes.ts

```diff
@@ -0,0 +1,174 @@
+import { uniqBy } from "lodash"
+import { IGatsbyState } from "../redux/types"
+
+const mapComponentsToStaticQueryHashes = (
+  staticQueryComponents: IGatsbyState["staticQueryComponents"]
+): Map<string, string> =>
+  Array.from(staticQueryComponents).reduce(
+    (map, [, { componentPath, hash }]) => {
+      map.set(componentPath, hash)
+      return map
+    },
+    new Map()
+  )
+
+/* This function takes the current Redux state and a compilation
+ * object from webpack and returns a map of unique templates
+ * to static queries included in each (as hashes).
+ *
+ * This isn't super straightforward because templates may include
+ * deep component trees with static queries present at any depth.
+ * This is why it is necessary to map templates to all their (user land and node_modules)
+ * dependencies first and then map those dependencies to known static queries.
+ *
+ * Also, Gatsby makes it possible to wrap an entire site or page with a layout
+ * or other component(s) via the wrapRootElement and wrapPageElement APIs. These must
+ * also be handled when computing static queries for a page.
+ *
+ * Let's go through the implementation step by step.
+ */
+export function mapTemplatesToStaticQueryHashes(
+  reduxState: IGatsbyState,
+  compilation
```

Comment:
add type here? I think you can get it from webpack

Reply:
Done